README.md
# dsci_310_group_11_pkg

This package contains functions used for generating dsci_310_group_11's report. The main functions inside this package is to 
split the data, generate pipeline for predictive models, conduct hyperparameter tunning and make graphs.

The main difference between our package to other similar functions is that the functions in our package is designed to improve the efficiency for analyzing the wine data. Using our pipeline generator as example, while directly using functions in sklearns can generate the same model, we simplifies this process by integrating repetitive steps into our functions. 

## Code Coverage Badge

[![codecov](https://codecov.io/gh/DSCI-310/dsci-310-group-11-pkg/branch/main/graph/badge.svg?token=e9QSz5vHox)](https://codecov.io/gh/DSCI-310/dsci-310-group-11-pkg)

## Installation

1. Clone the git repository onto your local machine using:
```
git clone https://github.com/DSCI-310/dsci-310-group-11-pkg.git
```

2. Navigate to the local repository through the terminal (this may differ based on your home directory setups):
```
cd dsci-310-group-11-pkg
```

3. Run pip to install the package locally (from [pypi](https://pypi.org/project/dsci-310-group-11-pkg/)):
```
pip install dsci-310-group-11-pkg
```

You should now be able to import the package into relevant projects and notebooks.

## Usage

The most basic usage of this package is to split the data into training or testing data:

```python
from dsci_310_group_11_pkg.preprocess import preprocessor

df # target dataframe
training_data = preprocessor(df, 0)
testing_data = preprocessor(df, 1)
```

One of the other usages of `dsci_310_group_11_pkg` is to generate pipeline for classification models as follows:

```python
from dsci_310_group_11_pkg.pipeline import pipe_build

x = X_training_data # training features data
y = Y_training data # training label data
model_type = 'lr' # types of model

model = pipe_build(model_type, x, y)

```
Another usage is to conduct hyperparameter tuning and return it as dataframe for different model as follows:

```python
from dsci_310_group_11_pkg.optimizer import optimize

x = X_training_data # training features data
y = Y_training data # training label data
model_name = 'lr' # types of model

tuning_result = optimize(model_name, x, y)
```

This package can also be used to generate different graphs for the analysis report, for example: 

```python
from dsci_310_group_11_pkg.grapher import correlation_table

ctb = correlation_table(df) # this generate the correlation table for dataframe df
```
Other functions and more detailed usage can be found in docs/example or in the [website documentation of the package](https://dsci-310-group11-pkg.readthedocs.io/en/latest/index.html).

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`dsci_310_group_11_pkg` was created by DSCI_310_Group_11. It is licensed under the terms of the MIT license.

## Credits

`dsci_310_group_11_pkg` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (06/04/2023)

- First release of `dsci_310_group_11_pkg`!

test/test_grapher.py
import numpy as np
import pandas as pd
from dsci_310_group_11_pkg.grapher import correlation_table, bar_chart, vis_tree, compare_scores, show_coefficients, show_correct
import altair
from matplotlib import pyplot as plt
import pytest
from sklearn.dummy import DummyClassifier

# dummy data

dummy = pd.read_csv('docs/data/winequality-red.csv', sep=',')[0:10]
dummy_X = dummy.drop(columns = ['quality'])
dummy_y = dummy['quality']
pipe = DummyClassifier(strategy='most_frequent').fit(dummy_X, dummy_y)



# test whether the graphing function returns the correct types of Charts or output
def test_correlation_table():
    # DESCRIPTION: Compare the type of chart returned by correlation_table is the expected type
    # ACTION: use is_json to compare the type of chart
    # RETURNS: return errors message if the type of chart is not the expected type, return none if the test pass

    assert type(correlation_table(dummy)) == altair.vegalite.v4.api.Chart
    
def test_bar_chart():
    # DESCRIPTION: Compare the type of chart returned by bar_chart is the expected type
    # ACTION: use is_json to compare the type of chart
    # RETURNS: return errors message if the type of chart is not the expected type, return none if the test pass
    
    assert type(bar_chart(dummy)) == altair.vegalite.v4.api.Chart
    
def test_vis_tree():
    # DESCRIPTION: Compare the type of chart returned by vis_tree is the expected type
    # ACTION: use assert and == function to compare the type of chart
    # RETURNS: return errors message if the type of chart is not the expected type, return none if the test pass
    
    assert vis_tree(dummy_X, dummy_y) is None
    plt.close()
    
def test_compare_scores():
    # DESCRIPTION: Compare the type of chart returned by compare_scores is the expected type
    # ACTION: use is_json function to compare the type of chart
    # RETURNS: return errors message if the type of chart is not the expected type, return none if the test pass
    
    assert type(compare_scores([1,2,3,4,5])) == altair.vegalite.v4.api.Chart

    
    
# test whether this function returns the correct number of True prediction
def test_show_correct():
    # DESCRIPTION: Compare the the number of true cases generated by show_correct to the real answer
    # ACTION: use assert and == function to compare the value of true cases
    # RETURNS: return errors message if number of true cases not equals the acutal true cases, return none if the test pass
    assert show_correct(pipe, dummy_X, dummy_y)[True] == 7

test/test_optimize.py
import numpy as np
import pandas as pd
from sklearn.datasets import make_moons
from dsci_310_group_11_pkg.optimize import hp_optimizer
import pytest



# dummy data
X_train, y_train = make_moons(n_samples = 100)


# test whether the function returns hyperparameter and validation score
def test_hp_optimizer():
    # DESCRIPTION: Compare the columns names of the dataframe that generated by the hp_optimizer function is the same as the expected column names
    # ACTION: Use assert and == function to compare the column names
    # RETURNS: Return errors message if the column names are not the same, return none if the test pass

    actual_lr = hp_optimizer("lr", X_train, y_train)
    assert list(actual_lr.columns) == ['C', 'mean_train_scores', 'mean_cv_scores']
    actual_dtc = hp_optimizer('dtc', X_train, y_train)
    assert list(actual_dtc.columns) == ['max_depth', 'mean_cv_score']
    actual_svm = hp_optimizer('svm', X_train, y_train)

test/test_pipeline.py
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import classification_report
from sklearn.dummy import DummyClassifier
from sklearn.naive_bayes import GaussianNB

from dsci_310_group_11_pkg.pipeline import pipe_build
from sklearn.datasets import make_moons
import pytest


## making data for test

X_train, y_train = make_moons(n_samples = 100)
X_test, y_test = make_moons(n_samples = 50)

## test for pipe_build

# DESCRIPTION: Compare the accruacy of expected models and actual models created by pipe_build, each case will be checked by a seperate test
# ACTION: use assert and score function to compare different models' accuracy 
# RETURNS: return errors message if the accuracy is not the same, return none if the test pass

def test_pipe_build_dummy():    
    expected = DummyClassifier(strategy='most_frequent').fit(X_train, y_train)
    actual = pipe_build('dummy', X_train, y_train)
    # assert actual == expected, \ ### I don't know why it always are false
    assert actual.score(X_test, y_test) == expected.score(X_test, y_test),\
            "Wrong DummyClassifier"
    
def test_pipe_build_lr():
    expected = make_pipeline(
                    StandardScaler(),
                    LogisticRegression(C=0.01, class_weight='balanced', random_state=1234)).fit(X_train, y_train)
    actual = pipe_build('lr', X_train, y_train)
    assert actual.score(X_test, y_test) == expected.score(X_test, y_test), \
            "Wrong logistic regressor"
    

def test_pipe_build_svm():
    expected = make_pipeline(
                StandardScaler(),
                SVC(C=1, gamma=0.1, class_weight='balanced', random_state=1234)).fit(X_train, y_train)
    actual = pipe_build('svm', X_train, y_train)
    assert actual.score(X_test, y_test) == expected.score(X_test, y_test),\
            "Wrong SVM model"
    
def test_pipe_build_dtc():
    expected = DecisionTreeClassifier(max_depth=7,
                             random_state=1234,
                             class_weight='balanced').fit(X_train, y_train)
    actual = pipe_build('dtc', X_train, y_train)
    assert actual.score(X_test, y_test) == expected.score(X_test, y_test),\
            "Wrong Decision tree model"
    
def test_pipe_build_bayes():
    expected = GaussianNB().fit(X_train, y_train)
    actual = pipe_build('bayes', X_train, y_train)
    assert actual.score(X_test, y_test) == expected.score(X_test, y_test),\
            "Wrong Bayes model"

src/dsci_310_group_11_pkg/grapher.py
import altair as alt
import pandas as pd
from matplotlib import pyplot as plt
from sklearn.tree import DecisionTreeClassifier, plot_tree
import seaborn as sns
from sklearn.metrics import classification_report

def correlation_table(df):
    """DESCRIPTION: Displays a correlation table (correlation coefficient value
    of each variable to each other variable).
    
    INPUTS: df - A dataframe object containing prediction features.
    
    ACTION: Inputs a dataframe and displays the correlation coefficients in
    a square grid.
    
    RETURNS: The table as a display.
    """
    cor_data = (
        df.corr().stack().reset_index(
        )  # The stacking results in an index on the correlation values, we need the index as normal columns for Altair
        .rename(columns={
            0: 'correlation',
            'level_0': 'Variable 1',
            'level_1': 'Variable 2'
        }))
    cor_data['correlation_label'] = cor_data['correlation'].map(
        '{:.2f}'.format)  # Round to 2 decimal

    # This is the inter-related correlations of each chemical ingredient
    # Judging on the correlations we may be able to infer coefficient signage.

    table = alt.Chart(cor_data).mark_rect().encode(
        alt.X('Variable 1:O'),
        alt.Y('Variable 2:O'),
        alt.Color('correlation'),
        alt.Tooltip(['correlation_label']),
    ).interactive().properties(width=300, height=300)

    xy = table.to_json('corrtab.json')
    return xy

def bar_chart(df):
    """DESCRIPTION: Displays a simple bar chart of the count of the quality variable.
    
    ACTION: Inputs a dataframe and displays the bar chart.
    
    INPUTS: df - A dataframe object
    
    RETURNS: The bar chart as a display.
    
    TODO: 1. Modularize the variables that you can input into the chart
    2. Move from altair to matplotlib
    """
    x = alt.Chart(df).mark_bar().encode(alt.X('quality:O'),
                                        alt.Y('count()')).properties(width=200,
                                                                     height=100)
    xyy = x.to_json('bar.json')
    return xyy

def class_report(pipe, X_test, y_test):
    """DESCRIPTION: Displays a heatmap of the count of the predicted case.
    
    ACTION: Inputs a model, testing data and displays the heatmap.
    
    INPUTS: pipe - a model
            X_test - testing features data
            y_test - testing label data
    
    RETURNS: The heatmap as figure
    
    """
    clf_report = classification_report(y_test, pipe.predict(X_test), output_dict=True)
    report = sns.heatmap(pd.DataFrame(clf_report).iloc[:-1, :].T, annot=True, linewidth=.5, cmap="crest")
    fig = report.get_figure()
    return fig

def vis_tree(X_train, y_train):
    """DESCRIPTION: Displays a visual example of a decision tree for conceptual
    purposes. The max_depth variable is limited to 3 so that the visualization
    is interpretable.
    
    INPUTS: X_train - a dataframe object containing prediction features
            y_train - a series object containing target variables.
    
    ACTION: Inputs an X_train dataframe and y_train series and displays the 
    decision tree model and each of its chosen parameter splits.
    
    RETURNS: The decisision tree model as a display.
    """
    vistree = DecisionTreeClassifier(max_depth=3,
                              random_state=1234,
                              class_weight='balanced')

    model2 = vistree.fit(X_train, y_train)

    fig = plt.figure(figsize=(15, 5))
    plot_tree(vistree, feature_names=X_train.columns, class_names='target', filled=True, fontsize=7)

def compare_scores(lst):
    """DESCRIPTION: Displays a bar chart comparing the accuracy scores of each
    ML model in the 'lst' list.
    
    INPUTS: lst - a list of floats (accuracy scores) of each model.
    
    ACTION: Inputs a list (lst) of ML model accuracy scores, generates a
    dataframe named 'report' and turns this dataframe into a bar chart.
    
    RETURNS: The bar chart where the highlighted bar is the highest score.
    """
    
    cscores = lst

    report = pd.DataFrame()
    report = report.append(pd.DataFrame(
        [cscores], columns=['Baseline', 'LR', 'SVC', 'DT', 'NB']),
                            ignore_index=True)

    report.index = ['Score']
    report = report.T.reset_index()

    y = alt.Chart(report).mark_bar().encode(
    alt.X('Score:Q'),
    alt.Y('index:N'),
    color=alt.condition(
        alt.datum.Score == max(
            report['Score']),  # If the year is 1810 this test returns True,
        alt.value('red'),  # which sets the bar orange.
        alt.value(
            'steelblue')  # And if it's not true it sets the bar steelblue.
    )).properties(width=500, height=200).configure(background='lightgrey')

    y = y.to_json('scores.json')
    return y

def show_coefficients(pipe, X_train):
    """DESCRIPTION: Displays a dataframe with the coefficients of the Logistic
    Regression model.
    
    INPUTS: pipe - a pipeline object containing scikit-learn model transformers, and a scikit-learn model.
            X_train - a dataframe object containing prediction features.
    
    ACTION: Inputs a LogisticRegression model, and an X_train dataset.
    Names the pipe variables given the named_steps in the logistic regression
    in an array called 'flatten'. Returns the dataframe 'coeffs' with the
    model's features versus their coefficients. Sorts the values descendingly.
   
    RETURNS: The dataframe, sorted descending by coefficients value.

    Printing out coefficients of the regression model for values influencing the model.
    """
    flatten = pipe.named_steps["logisticregression"].coef_  # 2-D Array
    flatten = flatten.flatten()  # Converting 2-D Array to 1-D Array

    coeffs = pd.DataFrame(
        data={
            "features": X_train.columns,  # 1-D Array
            "coefficients": flatten,  # 1-D Array
        })

    # intercept = pipe.named_steps["logisticregression"].intercept_
    # print(f"The intercept is: {intercept}")
    return coeffs.sort_values('coefficients', ascending=False)

def show_correct(pipe, X_test, y_test):
    """ DESCRIPTION: Displays a dataframe with the True Positive + True Negative
    versus the False Positive + False Negative ratio of the classifier model.
    
    INPUTS: pipe - a pipeline object containing scikit-learn model transformers, and a scikit-learn model.
            X_test - a dataframe object containing prediction features.
            y_test - a series object containing target variables.
    
    ACTION: Inputs a model (pipe), and testing data; calls predict on the 
    test data and reports the correct classifications versus the incorrect
    classifications.
    
    RETURNS: A dataframe with the correct classifications versus incorrect
    classifications.
    """

    ax = pd.DataFrame(data={'actual': y_test, 'predicted': pipe.predict(X_test)})
    ax['correct'] = ax['actual'] == ax['predicted']

    return ax.correct.value_counts()

src/dsci_310_group_11_pkg/pipeline.py
# Import required sklearn packages for model construction
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.dummy import DummyClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier


def pipe_build(model, X_train, y_train):
    # DESCRIPTION: Generates specified ML model based on 'model' as a string
    # INPUTS: model - an input string, referencing the specific model that the function should run.
            # X_train - a dataframe object containing prediction features.
            # y_train - a series object containing target variables.
    # ACTION: Creates a pipeline with appropriate transformations (scaling, etc.)
    # and fits the specified model with optimized hyperparameters to the training data
    # RETURNS: the pipeline object (model) for evaluation in wineclassification.ipynb
    # TODO: Modularize hyperparameter values
    
    if model == 'dummy':
        baseline = DummyClassifier(strategy='most_frequent')
        pipe = baseline.fit(X_train, y_train)
        
    elif model == 'lr':
        pipe = make_pipeline(
                    StandardScaler(),
                    LogisticRegression(C=0.01, class_weight='balanced', random_state=1234))
        pipe.fit(X_train, y_train)

    elif model == 'svm':
        pipe = make_pipeline(
                StandardScaler(),
                SVC(C=1, gamma=0.1, class_weight='balanced', random_state=1234))
        pipe.fit(X_train, y_train)

    elif model == 'dtc':
        clf = DecisionTreeClassifier(max_depth=7,
                             random_state=1234,
                             class_weight='balanced')
        pipe = clf.fit(X_train, y_train)

    elif model == 'bayes':
        pipe = GaussianNB()
        pipe.fit(X_train, y_train)  

    return pipe


src/dsci_310_group_11_pkg/preprocess.py
import numpy as np
from sklearn.model_selection import train_test_split

def preprocessor(df, tort):
    # DESCRIPTION: Splits the dataset of the specified 'df' (dataframe) into
    # training and testing data, generates a 'target' variable for the ML
    # model to classify given the value of the quality of each example.
    # INPUTS: df - a dataframe object that contains the entirety of the dataset, for splitting.
            # tort - a binary value (0, 1) that specifies whether to return the train or test dataframe.
    # ACTION: Splits the dataset of the specified 'df' (dataframe) into
    # training and testing data, uses np.where to assign a 0 to the target
    # column of examples that have quality < 5, assigns a 1 to the target
    # column of examples that have quality > 5.
    # RETURNS: IF the function calls 0, then returns the training data, ELSE IF
    # function calls, returns the testing data
    # TODO: Modularize param_grid values
    train_df, test_df = train_test_split(df, test_size=0.30, random_state=123)
    train_df['target'] = np.where(train_df['quality'] > 5, 1, 0)
    test_df['target'] = np.where(test_df['quality'] > 5, 1, 0)

    if tort == 0:
        return train_df
    elif tort == 1:
        return test_df


src/dsci_310_group_11_pkg/optimize.py
# Import Required Packages
import numpy as np
import pandas as pd

from sklearn.model_selection import cross_val_score, cross_validate, train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

def hp_optimizer(model, X_train, y_train):
    # DESCRIPTION: Maximizes model accuracy based on tuning hyperparameters of each model (respective if statements)
    # ACTION: Loops over chosen hyperparemeters in a the scores_dict/param_grid dictionary
    # and appends the mean cross validation score of each hyperparameter to the scores_dict dictionary
    # INPUTS: model - an input string, referencing the specific model that the function should run.
            # X_train - a dataframe object containing prediction features.
            # y_train - a series object containing target variables.
    # RETURNS: Dataframe containing the mean cross validation scores of each hyperparameter value
    # TODO: Modularize param_grid values
    
    # Logistic Regression model needs to optimize the "C" value (a value of model regularization)
    # C: Regularization; penalty of wrongly classified examples
    # Creates a simple ML pipeline that scales the data and applies LogisticRegression
    if model == 'lr':
        scores_dict = {
        "C": 10.0**np.arange(-4, 6, 1),
        "mean_train_scores": list(),
        "mean_cv_scores": list(),
        }
        for C in scores_dict["C"]:
            pipe = make_pipeline(StandardScaler(),
                                 LogisticRegression(C=C, random_state=1234))
            scores = cross_validate(pipe, X_train, y_train, return_train_score=True)
            scores_dict["mean_train_scores"].append(scores["train_score"].mean())
            scores_dict["mean_cv_scores"].append(scores["test_score"].mean())

        results_df = pd.DataFrame(scores_dict)
        return results_df

    # Support Vector Machine model optimizes the 'C' and 'gamma' values.
    # C: Regularization; penalty of wrongly classified examples
    # Gamma: Strenghth of the influence of individual examples
    elif model == 'svm':
        param_grid = {
            "C": [0.001, 0.01, 0.1, 1, 10, 100],
            "gamma": [0.001, 0.01, 0.1, 1, 10, 100]
        }

        results_dict = {"C": [], "gamma": [], "mean_cv_score": []}

        for gamma in param_grid["gamma"]:
            for C in param_grid["C"]:  # for each combination of parameters, train an SVC
                pipe_svm = make_pipeline(StandardScaler(), SVC(gamma=gamma, C=C, class_weight='balanced', random_state=1234))
                scores = cross_val_score(pipe_svm, X_train, y_train, cv=5)
                mean_score = np.mean(scores)
            
                results_dict["C"].append(C)
                results_dict["gamma"].append(gamma)
                results_dict["mean_cv_score"].append(mean_score)


        results_df = pd.DataFrame(results_dict).head(10)
        return results_df

    # Decision Tree Classification optimizes the 'max depth' value
    # Max Depth: How many parameters (splits/decisions) to stop at
    elif model == 'dtc':

        param_grid = {"max_depth": np.arange(1, 20, 2)}

        results_dict = {"max_depth": [], "mean_cv_score": []}
        for depth in param_grid["max_depth"]:
            dtc = DecisionTreeClassifier(max_depth=depth, class_weight='balanced', random_state=1234)
            scores = cross_val_score(dtc, X_train, y_train)
            mean_score = np.mean(scores)

            results_dict["max_depth"].append(depth)
            results_dict["mean_cv_score"].append(mean_score)

        results_df = pd.DataFrame(results_dict)
        return results_df
    
    else:
        return 'Error: You need to input a proper model'
